dimod/generators/gates.py

```diff
@@ -249,109 +252,66 @@ def multiplication_circuit(nbit: int) -> BinaryQuadraticModel:
-                # gate. And since we only have one bit to add, we don't need an adder, have no
-                # carry out
-                andij = AND[i][j] = p[0]
-                gate = and_gate(ai, bj, andij)
-                bqm.update(gate)
-                continue
-
-            # we always need an AND gate
-            andij = AND[i][j] = 'and%s,%s' % (i, j)
-            gate = and_gate(ai, bj, andij)
-            bqm.update(gate)
-
-            # the number of inputs will determine the type of adder
-            inputs = [andij]
-
-            # determine if there is a carry in
-            if i - 1 in CARRY and j in CARRY[i - 1]:
-                inputs.append(CARRY[i - 1][j])
-
-            # determine if there is a sum in
-            if i - 1 in SUM and j + 1 in SUM[i - 1]:
-                inputs.append(SUM[i - 1][j + 1])
-
-            # ok, create adders if necessary
-            if len(inputs) == 1:
-                # we don't need an adder and we don't have a carry
-                SUM[i][j] = andij
-            elif len(inputs) == 2:
-                # we need a HALFADDER so we have a sum and a carry
-                if j == 0:
-                    sumij = SUM[i][j] = p[i]
-                else:
-                    sumij = SUM[i][j] = 'sum%d,%d' % (i, j)
-
-                carryij = CARRY[i][j] = 'carry%d,%d' % (i, j)
-                gate = halfadder_gate(inputs[0], inputs[1], sumij, carryij)
-                bqm.update(gate)
-            else:
-                assert len(inputs) == 3, 'unexpected number of inputs'
-                # we need a FULLADDER so we have a sum and a carry
-                if j == 0:
-                    sumij = SUM[i][j] = p[i]
-                else:
-                    sumij = SUM[i][j] = 'sum%d,%d' % (i, j)
-
-                carryij = CARRY[i][j] = 'carry%d,%d' % (i, j)
-                gate = fulladder_gate(inputs[0], inputs[1], inputs[2], sumij, carryij)
-                bqm.update(gate)
-
-    # now we have a final row of full adders
-    for col in range(nbit - 1):
-        inputs = [CARRY[nbit - 1][col], SUM[nbit - 1][col + 1]]
-
-        if col == 0:
-            sumout = p[nbit + col]
-            carryout = CARRY[nbit][col] = 'carry%d,%d' % (nbit, col)
-            gate = halfadder_gate(inputs[0], inputs[1], sumout, carryout)
-            bqm.update(gate)
-            continue
-
-        inputs.append(CARRY[nbit][col - 1])
-
-        sumout = p[nbit + col]
-        if col < nbit - 2:
-            carryout = CARRY[nbit][col] = 'carry%d,%d' % (nbit, col)
-        else:
-            carryout = p[2 * nbit - 1]
-
-        gate = fulladder_gate(inputs[0], inputs[1], inputs[2], sumout, carryout)
-        bqm.update(gate)
-
-    return bqm
+    a = {i: 'a%d' % i for i in range(num_multiplier_bits)}
+    b = {j: 'b%d' % j for j in range(num_multiplicand_bits)}
+    p = {k: 'p%d' % k for k in  range(num_multiplier_bits + num_multiplicand_bits)}
```

Comment:
Looks like you no longer use `p`. Honestly we probably don't need to make these in advance at all

Reply:
I can make them functions like AND, SUM & CARRY. This his how they were to start of with. I have no strong feelings about these. unlike AND, SUM & CARRY which were adding unnecessary complexity as mutating nested dicts.

Reply:
... or icould just inline them as string interpolations

Reply:
I was imagining the inline, but whatever works.

Reply:
Done